src/main/java/de/tum/in/www1/artemis/service/ExamService.java

```diff
@@ -115,4 +140,161 @@ public void delete(Long examId) {
     public Set<Exam> filterVisibleExams(Set<Exam> exams) {
         return exams.stream().filter(Exam::isVisibleToStudents).collect(Collectors.toSet());
     }
+
+    /**
+     * Generates the student exams randomly based on the exam configuration and the exercise groups
+     *
+     * @param examId        the id of the exam
+     * @return the list of student exams with their corresponding users
+     */
+    public List<StudentExam> generateStudentExams(Long examId) {
+        List<StudentExam> studentExams = new ArrayList<>();
+        SecureRandom random = new SecureRandom();
+
+        // Delete all existing student exams via orphan removal
+        Exam examWithExistingStudentExams = examRepository.findWithStudentExamsById(examId).get();
+        studentExamRepository.deleteInBatch(examWithExistingStudentExams.getStudentExams());
+
+        Exam exam = examRepository.findWithExercisesRegisteredUsersStudentExamsById(examId).get();
+
+        // Ensure that all exercise groups have at least one exercise
+        for (ExerciseGroup exerciseGroup : exam.getExerciseGroups()) {
+            if (exerciseGroup.getExercises().isEmpty()) {
+                throw new BadRequestAlertException("All exercise groups must have at least one exercise", "Exam", "artemisApp.exam.validation.atLeastOneExercisePerExerciseGroup");
+            }
+        }
+
+        // Check that numberOfExercisesInExam is set
+        if (exam.getNumberOfExercisesInExam() == null) {
+            throw new BadRequestAlertException("The number of exercises in the exam is not set.", "Exam", "artemisApp.exam.validation.numberOfExercisesInExamNotSet");
+        }
+
+        List<ExerciseGroup> exerciseGroups = exam.getExerciseGroups();
+
+        // Check that there are enough exercise groups
+        if (exerciseGroups.size() < exam.getNumberOfExercisesInExam()) {
+            throw new BadRequestAlertException("The number of exercise groups is too small", "Exam", "artemisApp.exam.validation.tooFewExerciseGroups");
+        }
+
+        long numberOfOptionalExercises = exam.getNumberOfExercisesInExam() - exerciseGroups.stream().filter(ExerciseGroup::getIsMandatory).count();
```

Comment:
Do we really need this to be a long?

Reply:
stream.count() returns a long, and I did not want to cast this to an int.
I don't think it makes a difference

Reply:
The implementation of `count()` for Java streams returns a long :-)